File: scripts/session/export.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'fileutils'
require 'time'
require 'optparse'

module Gmux
  module Scripts
    class SessionExport
      STORE_PATH = File.join(Dir.home, '.gmux', 'sessions.json')

      def initialize(options = {})
        @session_name = options[:session_name]
        @output_format = options[:format] || 'json'
        @output_file = options[:output_file]
        @all = options[:all] || false
        @verbose = options[:verbose] || false
      end

      def run
        puts "=== GMUX Session Export ==="
        puts

        sessions = load_sessions
        if sessions.empty?
          puts "No sessions found."
          return
        end

        if @all
          export_sessions(sessions)
        elsif @session_name
          session = sessions[@session_name]
          if session.nil?
            puts "Session '#{@session_name}' not found."
            return
          end
          export_session(@session_name, session)
        else
          puts "Specify a session name or use --all"
          return
        end
      end

      private

      def load_sessions
        return {} unless File.exist?(STORE_PATH)

        JSON.parse(File.read(STORE_PATH))
      rescue JSON::ParserError => e
        puts "Error parsing sessions file: #{e.message}"
        {}
      end

      def export_sessions(sessions)
        export_data = sessions.map { |name, session| export_session_data(name, session) }
        write_export(export_data)
      end

      def export_session(name, session)
        export_data = export_session_data(name, session)
        write_export(export_data)
      end

      def export_session_data(name, session)
        {
          name: name,
          branch: session['branchName'],
          worktree: session['worktreePath'],
          agent: session['agentCommand'],
          status: session['status'],
          started: session['startedAt'],
          exported_at: Time.now.iso8601
        }
      end

      def write_export(data)
        case @output_format
        when 'json'
          output = JSON.pretty_generate(data)
        when 'yaml'
          output = to_yaml(data)
        when 'toml'
          output = to_toml(data)
        else
          puts "Unknown format: #{@output_format}"
          return
        end

        if @output_file
          File.write(@output_file, output)
          puts "Exported to: #{@output_file}"
        else
          puts output
        end
      end

      def to_yaml(data)
        require 'yaml'
        data.to_yaml
      end

      def to_toml(data)
        lines = []
        Array(data).each do |item|
          lines << "[[session]]"
          item.each do |key, value|
            lines << "#{key} = \"#{value}\""
          end
          lines << ""
        end
        lines.join("\n")
      end
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: #{$PROGRAM_NAME} [options] [session_name]"

    opts.on('-a', '--all', "Export all sessions") do
      options[:all] = true
    end

    opts.on('-f', '--format FORMAT', %w[json yaml toml], "Output format (json, yaml, toml)") do |f|
      options[:format] = f
    end

    opts.on('-o', '--output FILE', "Write to file instead of stdout") do |o|
      options[:output_file] = o
    end

    opts.on('-v', '--verbose', "Show detailed output") do
      options[:verbose] = true
    end

    opts.on('-h', '--help', "Show this help") do
      puts opts
      exit
    end
  end.parse!

  options[:session_name] = ARGV.shift
  Gmux::Scripts::SessionExport.new(options).run
end

File: scripts/session/export.py
#!/usr/bin/env python3
"""GMUX Session Export - Export session configuration."""

import json
import sys
from argparse import ArgumentParser
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

STORE_PATH = Path.home() / ".gmux" / "sessions.json"


def _try_import_yaml():
    try:
        import yaml
        return yaml
    except ImportError:
        return None


def _try_import_toml():
    try:
        import tomllib
        return tomllib
    except ImportError:
        pass
    try:
        import tomli
        return tomli
    except ImportError:
        return None


class SessionExport:
    def __init__(self, options: dict):
        self.session_name = options.get("session_name")
        self.output_format = options.get("format", "json")
        self.output_file = options.get("output_file")
        self.all_sessions = options.get("all", False)
        self.verbose = options.get("verbose", False)

    def run(self):
        print("=== GMUX Session Export ===")
        print()

        sessions = self._load_sessions()
        if not sessions:
            print("No sessions found.")
            return

        if self.all_sessions:
            self._export_sessions(sessions)
        elif self.session_name:
            session = sessions.get(self.session_name)
            if session is None:
                print(f"Session '{self.session_name}' not found.")
                return
            self._export_session(self.session_name, session)
        else:
            print("Specify a session name or use --all")

    def _load_sessions(self) -> dict:
        if not STORE_PATH.exists():
            return {}
        try:
            return json.loads(STORE_PATH.read_text())
        except (json.JSONDecodeError, OSError) as e:
            print(f"Error parsing sessions file: {e}")
            return {}

    def _export_sessions(self, sessions: dict):
        data = [self._build_export(name, session) for name, session in sessions.items()]
        self._write(data)

    def _export_session(self, name: str, session: dict):
        data = self._build_export(name, session)
        self._write(data)

    def _build_export(self, name: str, session: dict) -> dict[str, Any]:
        return {
            "name": name,
            "branch": session.get("branchName"),
            "worktree": session.get("worktreePath"),
            "agent": session.get("agentCommand"),
            "status": session.get("status"),
            "started": session.get("startedAt"),
            "exported_at": datetime.now(timezone.utc).isoformat(),
        }

    def _write(self, data: Any):
        match self.output_format:
            case "json":
                output = json.dumps(data, indent=2)
            case "yaml":
                yaml = _try_import_yaml()
                if yaml is None:
                    print("Error: PyYAML is required for YAML export. Install with: pip install pyyaml")
                    return
                output = yaml.dump(data, default_flow_style=False)
            case "toml":
                output = self._to_toml(data)
            case _:
                print(f"Unknown format: {self.output_format}")
                return

        if self.output_file:
            Path(self.output_file).write_text(output)
            print(f"Exported to: {self.output_file}")
        else:
            print(output)

    def _to_toml(self, data: Any) -> str:
        items = data if isinstance(data, list) else [data]
        lines: list[str] = []
        for item in items:
            lines.append("[[session]]")
            for key, value in item.items():
                value_str = str(value) if value is not None else ""
                lines.append(f'{key} = "{value_str}"')
            lines.append("")
        return "\n".join(lines)


def main():
    parser = ArgumentParser(description="Export GMUX session configuration")
    parser.add_argument(
        "session_name",
        nargs="?",
        help="Session name to export",
    )
    parser.add_argument(
        "-a", "--all",
        action="store_true",
        dest="all",
        help="Export all sessions",
    )
    parser.add_argument(
        "-f", "--format",
        choices=["json", "yaml", "toml"],
        default="json",
        help="Output format (default: json)",
    )
    parser.add_argument(
        "-o", "--output",
        dest="output_file",
        help="Write to file instead of stdout",
    )
    parser.add_argument(
        "-v", "--verbose",
        action="store_true",
        help="Show detailed output",
    )

    args = parser.parse_args()
    SessionExport(vars(args)).run()


if __name__ == "__main__":
    main()
